lib/srv/desktop/tdp/proto.go

```diff
@@ -443,6 +453,155 @@ func decodeClipboardData(in peekReader, maxLen uint32) (ClipboardData, error) {
 	return ClipboardData(b), nil
 }
 
+const maxMFADataLength = 1024 * 1024
+
+type MFA struct {
+	// Type should be one of defaults.WebsocketU2FChallenge or defaults.WebsocketWebauthnChallenge
+	Type byte
+	// MFAAuthenticateChallenge is the challenge we send to the client.
+	// Used for messages from Teleport to the user's browser.
+	*auth.MFAAuthenticateChallenge
+	// MFAAuthenticateResponse is the response to the MFA challenge,
+	// sent from the browser to Teleport.
+	*authproto.MFAAuthenticateResponse
+}
+
+func (m MFA) Encode() ([]byte, error) {
+	buf := new(bytes.Buffer)
+	buf.WriteByte(byte(TypeMFA))
+	buf.WriteByte(m.Type)
+	var buff []byte
+	var err error
+
+	log.Debug("MYDEBUG  ->>>> send buff 111 ")
+	if m.MFAAuthenticateChallenge != nil {
+		buff, err = json.Marshal(m.MFAAuthenticateChallenge)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+	} else if m.MFAAuthenticateResponse != nil {
+		switch t := m.MFAAuthenticateResponse.Response.(type) {
+		case *authproto.MFAAuthenticateResponse_U2F:
+			buff, err = json.Marshal(m.MFAAuthenticateResponse.GetU2F())
+			if err != nil {
+				return nil, trace.Wrap(err)
+			}
+		case *authproto.MFAAuthenticateResponse_Webauthn:
+			buff, err = json.Marshal(m.MFAAuthenticateResponse.GetWebauthn())
+			if err != nil {
+				return nil, trace.Wrap(err)
+			}
+		default:
+			return nil, trace.BadParameter("unsupported type %T", t)
+		}
+	} else {
+		return nil, trace.BadParameter("got nil MFAAuthenticateChallenge and MFAAuthenticateResponse fields")
+	}
+
+	if len(buff) > maxMFADataLength {
+		return nil, trace.BadParameter("mfa challenge data exceeds maximum length")
+	}
+	binary.Write(buf, binary.BigEndian, uint32(len(buff)))
+	log.Debug("MYDEBUG  ->>>> send buff ", string(buff))
+	buf.Write(buff)
+	return buf.Bytes(), nil
+}
+
+func DecodeMFA(in peekReader) (*MFA, error) {
+	t, err := in.ReadByte()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	if t != byte(TypeMFA) {
+		return nil, trace.BadParameter("got message type %v, expected TypeMFAJson(%v)", t, TypeMFA)
```

Comment:
```suggestion
		return nil, trace.BadParameter("got message type %v, expected TypeMFA(%v)", t, TypeMFA)
```

my rename didn't catch this one